### ISSUE TITLE
Error: *** gave up after 3 retries

### ISSUE THREAD

**Issue by sinhayash** [USC-NSL/miniNExT#9]
In a vm having newly installed Ubuntu 14.04, I ran the following commands

```
sudo apt-get install mininet
sudo git clone https://github.com/USC-NSL/miniNExT
cd miniNExT/
sudo apt-get install 'make deps'
sudo make install
cd ..
sudo apt-get install quagga quagga-doc
sudo python  miniNExT/examples/quagga-ixp/start.py 
```

OS: Ubuntu 14.04
Mininet version: 2.1.0
OvS version: 2.0.2

But I am receiving the following error.

```
** Creating Quagga network topology
** Starting the network
** Using Mininet Extended (MiniNExT) Handler
*** Creating network
*** Adding controller
*** Adding hosts:
a1 b1 c1 c2 d1 rs 
*** Adding switches:
fabric-sw1 
*** Adding links:
*** gave up after 3 retries
```

Am I missing something?

**Comment by edgar-costa** [comment]
I have the same problem, however only when I upgrade the linux kernel.  With kernel 3.19 it works for me, but once I upgrade to kernel 4.x.y  I start getting that error. 

If you try with mininet 2.2.1 it does work, but that mininet version is not supported by miniNExT (and I am not sure if it would be difficult to make it work). If you run mininet with "debug" option you can see what is going on, and maybe you are able to solve it. To do that edit the file start.py change setLogLevel('info') and put setLogLevel('debug').

If I find a way to fix that problem I will write it here. Tell us if you find a solution.

Edgar

**Comment by mjnovice** [comment]
I was facing the same issue, and for me the problem behind was http://stackoverflow.com/questions/24932172/what-length-can-a-network-interface-name-have . The length of the network interface name cannot exceed 15/16 bytes, due to which the creation of interface was failing, and hence was unable to form a link ( This does not seem to be the case here, as the host names are pretty small ).

This is silently failing because the mininet 2.1.0 (around which MiniNExT is built) https://github.com/mininet/mininet/blob/2.1.0/mininet/node.py ( line 355) is called without the printError option https://github.com/mininet/mininet/blob/2.1.0/mininet/util.py ( line 194). To prevent this from silently failing, either one could implement this function in the subclassed Host object adding the printError option in the call made at line 355.

**Comment by mjnovice** [comment]
@sinhayash, @edgarcosta92 could you tell, if the above is the case ?

**Comment by sinhayash** [comment]
Not sure, but I don't think I have used any interface 16 bytes long while running the above experiment, so it seems it is less likely to be the case

**Comment by edgar-costa** [comment]
My case its after a kernel upgrade from 3.x to 4.x. Right now I am trying to write my own miniNext so it works with my new kernel and mininet 2.2.1

**Comment by mjnovice** [comment]
Yeah true, but my point is that it should not be failing silently. If the printError option is enabled, it would provide better visibility into what exactly the problem is.

**Comment by bschlinker** [comment]
@mjnovice To be clear, all of the code that controls setting up these network interfaces is in Mininet, not MiniNExT. I agree that it would be good to not 'hide' the errors in these cases, but that's not on MiniNExT's side.

**Comment by LuizaNacshon** [comment]
Hey @sinhayash and all ,

I'm facing the same problem as you did with mininet  2.1.0p1 on Ubuntu 16.04, 4.4.0-31-generic
When I run the command 
$ sudo mn --mac --switch=ovsk,protocols=OpenFlow13 --controller=remote,ip=127.0.0.1,port=6653 --topo=tree,3

I got :
**\* Creating network
**\* Adding controller
**\* Adding hosts:
h1 h2 h3 h4 h5 h6 h7 h8 
**\* Adding switches:
s1 s2 s3 s4 s5 s6 s7 
**\* Adding links:
****\* gave up after 3 retries**

Have you solved the problem ? is there any proper solution there?

Thanks,
Luiza

**Comment by chandanD4** [comment]
Clear the older instance of mininet ..  This issue could be due to mismatch .

execute following and give try . it should work

sudo rm -rf /usr/local/bin/mn /usr/local/bin/mnexec \
    /usr/local/lib/python_/_/_mininet_ \
    /usr/local/bin/ovs-\* /usr/local/sbin/ovs-*

**Comment by demurgos** [comment]
I stil experience this issue. The `rm` did not solve my problem. Could one of you clarify the issue?

**Comment by jrtmendoza** [comment]
> I stil experience this issue. The rm did not solve my problem. Could one of you clarify the issue?

 

I'll give it a shot. Here it goes. 

The error is due to the network interface name exceeding 15/16 bytes. The workaround I did is to rename the switch to something shorter. Instead of 'fabric-sw1', I just used 'sw1'. You can validate this by setting the verbosity of the mininet script to "debug". Edit start.py and change the parameter of setLogLevel to 'debug' then launch the script. That should show which point in your code you're running into trouble.

**Comment by vladasov** [comment]
> My case its after a kernel upgrade from 3.x to 4.x. Right now I am trying to write my own miniNext so it works with my new kernel and mininet 2.2.1

@edgar-costa Did you manage to write your own version of miniNExT that works with kernel 4.x and supports some version of mininet different from 2.1.0? Can you please give me any feedback on what you have done (or tried) and was it successful?

Thanks, Vladimir
